### PR TITLE
Enable Domain module fully on runtime5

### PR DESCRIPTION
**Pull request by mshinwell**
This enables the `Domain` module fully on runtime5.  One consequence of this is that the domain-local state (`Domain.DLS` ) module changes under runtime5 to the upstream runtime5 implementation, rather than what we are using for runtime4.

Not yet ready for review, will mark as draft for now.  (The diff must be viewed in a whitespace-insensitive mode, though.)

Based on #2205 to avoid conflicts

### CODE REVIEW

**Comment by mshinwell** [comment]
@Ekdohibs could you please review this?  For the stdlib `Domain` module, probably best to compare against 5.1.1 upstream.

File: ocaml/middle_end/semantics_of_primitives.ml [hunk 1]
```diff
@@ -160,7 +160,7 @@ let for_primitive (prim : Clambda_primitives.primitive) =
   | Pget_header _ -> No_effects, No_coeffects
   | Pdls_get ->
       (* only read *)
-      No_effects, No_coeffects
```
Comment (Ekdohibs) [inline ocaml/middle_end/semantics_of_primitives.ml:163]: Why was this changed?

Comment (mshinwell) [reply ocaml/middle_end/semantics_of_primitives.ml:163]: Hmm, well I think this operation has coeffects - it's a memory load?

Comment (Ekdohibs) [reply ocaml/middle_end/semantics_of_primitives.ml:163]: I checked upstream and it's `No_coeffects` there; I think it's correct because while it's a memory load, it's never to a location that is written to.

**Comment by mshinwell** [comment]
Adding the `to_upstream` tag for the change in `ocaml/middle_end/semantics_of_primitives.ml`
